### PR TITLE
Expose the number of gathered peers in the optimizer

### PR DESCRIPTION
**Pull request by IAL32**
This PR exposes the number of peers participating in the last averaging round to `hivemind.Optimizer`.

### CODE REVIEW

File: hivemind/optim/optimizer.py [hunk 1]
```diff
@@ -344,6 +344,13 @@ def _compute_schema_hash(self) -> int:
     def is_alive(self) -> bool:
         return self.state_averager.is_alive()
 
+    @property
+    def gathered_peers(self) -> int:
```
Comment (borzunov) [inline hivemind/optim/optimizer.py:348]: @justheuristic Is it a good idea to expose this or this info is already exposed somewhere?

Also, I'd expect a property named `gathered_peers` to return a collection of peers (e.g., peer IDs), not just the count of them. An alternative is to rename this property to something like `num_gathered_peers`.

**Review by borzunov — CHANGES_REQUESTED** [review]
Thanks for the PR! I'll double-check with my colleagues that we indeed don't expose this info exposed somewhere else. Meanwhile, I've left a minor comment regarding the naming/semantics.

File: hivemind/optim/optimizer.py (hunk 1 above)
Comment (justheuristic) [reply hivemind/optim/optimizer.py:348]: Hi, @IAL32 !

I recognize that it is convenient to keep track of group size, but merging this specific code into hivemind may introduce confusion in some modes. If use_local_updates=False, optimizer will have two different averaging rounds (for state and for gradients), and it is not clear which of them we should report. Furthermore, if [moshpit](https://arxiv.org/abs/2103.03239) is enabled, optimizer can run multiple concurrent averaging rounds.

If you keep track of it in your experiments, you can subclass hivemind.Optimizer as such:

```python

class MyOptimizer(hivemind.Optimizer)
    @property
    def gathered_peers(self) -> int:
        return len(self.state_averager.gathered_peers)

    def _make_state_averager(self, **kwargs) -> TrainingStateAverager:
        # https://github.com/learning-at-home/hivemind/blob/master/hivemind/optim/optimizer.py#L282-L297
        return MyModifiedStateAverager(see_base_funciton_code(**kwargs))

class MyModifiedStateAverager(TrainingStateAverager):
    gathered_peers = ...

```

**Comment by borzunov** [comment]
@IAL32 By the way, if you are just looking for a way to fetch the current number of active collaborators, you can use the `optimizer.tracker.global_progress.num_peers` property. It is periodically updated basing on the DHT data.

For the reference, see the definition of the [GlobalTrainingProgress](https://github.com/learning-at-home/hivemind/blob/master/hivemind/optim/progress_tracker.py#L25) structure.

**Comment by borzunov** [comment]
Unfortunately, we're not ready to merge this PR in its current form. We propose you to use one of the workarounds discussed above: (a) fetch the number of active peers from the `GlobalTrainingProgress` class or (b) inherit `hivemind.Optimizer` and `TrainingStateAverager` to expose the `gathered_peers` property without changing the code of the library.

Feel free to reach us if you think we've missed something or if you need any help.

**Pull request closed without merging**